### DISCUSSION TITLE
URDF vs ProtoSpawner

### DISCUSSION THREAD

**Discussion by willgraham345** [cyberbotics/webots_ros2#860] [Q&A] [answered]
Had a quick question regarding #818. Has this PROTOSpawner feature been implemented, and is there any documentation regarding how it should be used?  I'm deciding between using the new feature, and URDF spawning.

**Comment by omichel** [comment, marked as answer]
It has not yet be fully implemented, but only partially with `spawn_node_from_string` to which you can pass a PROTO node. Using this method will be very close to the new feature and very easy to update when the new feature becomes available whereas the URDF spawner is going to be deprecated and will be harder to adapt to the new API. So, I would strongly recommend to go for the ProtoSpawner, on which you will also have a better control for setting Webots specific parameters.

**Reply by willgraham345** [reply]
Great! I'll plan to use the spawn_node_from_string. Is there example or reference code located within the webots_ros2 library? If so, is that compatible with ros2 humble?

**Reply by omichel** [reply]
I don't believe there is an example about it in this repo. However searching GitHub gives some [interesting results](https://github.com/search?q=spawn_node_from_string+&type=code). Maybe you can ask @TBreeze98 who implemented this feature in #539.

**Reply by willgraham345** [reply]
Those searches you mentioned are quickly murky. Most are shell scripts or source code of Webots itself. From what I can tell, the function this relies on the [importMFNodeFromString](https://cyberbotics.com/doc/reference/supervisor?tab-language=python#wb_supervisor_field_import_mf_node_from_string). I was able to call the service and create an empty Robot field, but wasn't able to fill the simulation with anything else. Is there more support regarding this function?

I don't know if there's an easy way to contact @TBreeze98. Their GitHub profile does not have any clear indicators on an email address and it appears that the account hasn't been active for some time. 

It appears that the URDF spawner is the best option for now, until more documentation can be generated.

**Comment by omichel** [comment]
> I was able to call the service and create an empty Robot field, but wasn't able to fill the simulation with anything else. Is there more support regarding this function?

If you replace the Robot node with a PROTO node, e.g., for example a Spot, it should work and insert the Spot robot in your world.
Don't forget to declare it in the world file as an IMPORTABLE PROTO:
```js
IMPORTABLE EXTERNPROTO "../path/to/my/protos/Spot.proto"
```
